lib/collection/src/shards/local_shard/query.rs

```diff
@@ -92,41 +147,38 @@ impl LocalShard {
         'shard: 'query,
     {
         async move {
-            let mut sources: Vec<Vec<ScoredPoint>> = Vec::with_capacity(prefetch.sources.len());
-
-            for source in prefetch.sources.into_iter() {
-                let vec: Vec<Vec<ScoredPoint>> = match source {
-                    PrefetchSource::SearchesIdx(idx) => {
-                        // TODO(universal-query): don't clone, by using something like a hashmap instead of a vec
-                        let scored_searches = core_results.get(idx).cloned().unwrap_or_default();
-                        vec![scored_searches]
-                    }
-                    PrefetchSource::ScrollsIdx(idx) => {
-                        // TODO(universal-query): don't clone, by using something like a hashmap instead of a vec
-                        let scrolled = scrolls.get(idx).cloned().unwrap_or_default();
-                        vec![scrolled]
-                    }
+            let mut search_indices = Vec::new();
+            let mut scroll_indices = Vec::new();
+
+            let mut merged_list = Vec::new();
+
+            for source in merge_plan.sources.into_iter() {
+                match source {
+                    PrefetchSource::SearchesIdx(idx) => search_indices.push(idx),
+                    PrefetchSource::ScrollsIdx(idx) => scroll_indices.push(idx),
                     PrefetchSource::Prefetch(prefetch) => {
-                        self.recurse_prefetch(
-                            prefetch,
-                            core_results,
-                            scrolls,
-                            search_runtime_handle,
-                            timeout,
-                            depth + 1,
-                        )
-                        .await?
+                        let merged = self
+                            .recurse_prefetch(
+                                prefetch,
+                                prefetch_holder,
+                                search_runtime_handle,
+                                timeout,
+                                depth + 1,
+                            )
+                            .await?;
+                        merged_list.extend(merged);
                     }
-                };
-                sources.extend(vec);
+                }
             }
 
-            if depth == 0 && prefetch.merge.rescore == Some(ScoringQuery::Fusion(Fusion::Rrf)) {
+            let sources = prefetch_holder.iter_sources(search_indices, scroll_indices, merged_list);
+
+            if depth == 0 && merge_plan.merge.rescore == Some(ScoringQuery::Fusion(Fusion::Rrf)) {
```

Comment:
```suggestion
            if depth == 0 && merge_plan.merge.rescore == Some(ScoringQuery::Fusion(_)) {
```

Reply:
Maybe better to turn into an exhaustive match, or does every fusion need all intermediate results?